gobblin-runtime/src/main/java/org/apache/gobblin/runtime/fork/AsynchronousFork.java

```diff
@@ -88,18 +90,25 @@ protected void processRecords() throws IOException, DataConversionException {
 
   @Override
   protected boolean putRecordImpl(Object record) throws InterruptedException {
+    if (record != Fork.SHUTDOWN_RECORD) {
+      this.recordsToProcess++;
+    }
     return this.recordQueue.put(record);
   }
 
   boolean processRecord() throws IOException, DataConversionException {
     try {
       Object record = this.recordQueue.get();
       if (record == null || record == Fork.SHUTDOWN_RECORD) {
+        if (record == null) {
+          log.info("Fork received a null record. isParentTaskDone=" + this.isParentTaskDone());
```

Comment:
Maybe doesn't have to be done in this PR, but we should explore emitting metrics on null records